Library/Homebrew/os.rb

```diff
@@ -45,8 +45,16 @@ def self.kernel_name
 
   ::OS_VERSION = ENV.fetch("HOMEBREW_OS_VERSION").freeze
 
-  CI_GLIBC_VERSION = "2.23"
-  CI_OS_VERSION = "Ubuntu 16.04"
+  LINUX_CI_OS_VERSION = "Ubuntu 16.04"
+  LINUX_GLIBC_FORMULA_VERSION = "2.23"
+  LINUX_GLIBC_CI_VERSION = "2.23"
+  LINUX_GCC_CI_VERSION = "5.0"
+  LINUX_PREFERRED_GCC_FORMULA = "gcc@5"
+
+  # Ubuntu 22.04 (see Linux-CI.md)
+  LINUX_GLIBC_NEXT_CI_VERSION = "2.35"
+  # LINUX_GCC_CI_VERSION = "11.0"
+  # LINUX_PREFERRED_GCC_FORMULA = "gcc@11"
```

Comment:
We should just make this the newest GCC, tbh. The newest runtime libraries should work in place of the slightly older ones. But that can be done separately.

Reply:
`CompilerSelector.preferred_gcc` needs to be the same as the version of GCC in the CI bottling distribution to ensure that formulae that do not have an explicit dependency on `gcc` are built and bottled using the default preferred version of GCC provided by the bottling distribution. If the brewed `gcc` happens to be installed by a transitive dependency, you don't want to build the bottle of other formulae without an explicit `gcc` dependency using the brewed `gcc`.

Reply:
> If the brewed `gcc` happens to be installed by a transitive dependency, you don't want to build the bottle of other formulae without an explicit `gcc` dependency using the brewed `gcc`.

Why don't you want that?

Reply:
If you're concerned about linkage to Homebrew GCC runtime libraries, we recently changed this so that everything prefers those libraries to the system's whenever Homebrew GCC is installed: https://github.com/Homebrew/brew/pull/13659

Reply:
@sjackman Nudge on this as it seems an important part of the imminent migration.

Reply:
On the most recent Ubuntu LTS system it's a better user experience if you can `brew install foobar` and use both the host's Glibc and GCC. If we set the minimum required version of GCC to 12, then on Ubuntu if you `brew install foobar` will first need to install `gcc` and its dependencies. One example use case of where that's inconvenient is using Homebrew on Linux for CI. You want to say `brew install jq` or something small, and it starts downloading and installing `gcc`, and it does that for every CI run (unless you go to the effort of setting up caching). I'm fine with older versions of Ubuntu LTS requiring a newer brewed toolchain, but the toolchain of the most recent version of Ubuntu LTS shouldn't require that.

A different proposal is that bottles are built with GCC 11, and if the host's compiler is older than GCC 11, it downloads and install the newest available brewed GCC, like GCC 12. That's possible, but if we want to go that route, we need to separate these two constants are ensure that bottles are built with the host's GCC even when the brewed GCC is installed (unless the brewed GCC is explicitly requested).

Reply:
> A different proposal is that bottles are built with GCC 11, and if the host's compiler is older than GCC 11, it downloads and install the newest available brewed GCC, like GCC 12.

Yes, that was what I was suggesting. Apologies for the lack of clarity here.

Reply:
IMO we should continue to use the same version of GCC as the host when it's older. It's much newer now and we can continue to have it updated by updating to future Ubuntu LTS versions more regularly.

Reply:
I thought my document here https://github.com/Homebrew/brew/blob/master/docs/Linux-CI.md#past-and-next-versions made clear that we decided to use gcc 11 (with Ubuntu 22.04). But maybe it's not clear and we should reword it / add a section about the gcc choice.

Reply:
@iMichka I think we're in (almost 😉) complete agreement there so: it's fine.